evm_arithmetization/src/cpu/kernel/interpreter.rs

```diff
@@ -382,117 +341,54 @@ impl<'a, F: Field> Interpreter<'a, F> {
         self.set_memory_multi_addresses(&block_hashes_fields);
     }
 
-    fn checkpoint(&self) -> InterpreterCheckpoint {
-        let registers = InterpreterRegistersState {
-            kernel_mode: self.is_kernel(),
-            context: self.context(),
-            registers: self.generation_state.registers,
-        };
-        InterpreterCheckpoint {
-            registers,
-            mem_len: self.memops.len(),
-        }
-    }
-
-    fn roll_memory_back(&mut self, len: usize) -> Result<(), ProgramError> {
-        // We roll the memory back until `memops` reaches length `len`.
-        debug_assert!(self.memops.len() >= len);
-        while self.memops.len() > len {
-            if let Some(op) = self.memops.pop() {
-                match op {
-                    InterpreterMemOpKind::Push(context) => {
-                        self.generation_state.memory.contexts[context].segments
-                            [Segment::Stack.unscale()]
-                        .content
-                        .pop()
-                        .ok_or(ProgramError::StackUnderflow)?;
-                    }
-                    InterpreterMemOpKind::Pop(value, context) => {
-                        self.generation_state.memory.contexts[context].segments
-                            [Segment::Stack.unscale()]
-                        .content
-                        .push(value);
-                    }
-                    InterpreterMemOpKind::Write(value, context, segment, offset) => {
-                        self.generation_state.memory.contexts[context].segments
-                            [segment >> SEGMENT_SCALING_FACTOR] // we need to unscale the segment value
-                            .content[offset] = value;
+    /// Applies all memory operations since the last checkpoint.
+    pub(crate) fn apply_memops(&mut self) -> Result<(), anyhow::Error> {
+        for memop in &self.generation_state.traces.memory_ops {
+            let &MemoryOp {
+                kind,
+                address,
+                value,
+                ..
+            } = memop;
+            match kind {
+                MemoryOpKind::Read => {
+                    if self.generation_state.memory.get_option(address).is_none() {
+                        if !self
+                            .preinitialized_segments
+                            .contains_key(&Segment::all()[address.segment])
+                        {
+                            assert_eq!(
```

Comment:
This will crash the process on failure. You sure that's what we want to happen here? If one looks at the method signature you would think that this method will return some sort of Error when things go wrong, but I don't see an error being returned anywhere. What's the thinking there? In general, if crashing is the best option, then the docs for the method should clearly state when it will crash.